x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,680 @@
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsBindingOnBucket)
+	store.Set(types.GetGVGsBindingOnBucketKey(gvgsBindingOnBucket.BucketId), bz)
+}
+
+func (k Keeper) SetGVGFamily(ctx sdk.Context, primarySpID uint32, gvgFamily *types.GlobalVirtualGroupFamily) {
+
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgFamily)
+	store.Set(types.GetGVGFamilyKey(primarySpID, gvgFamily.Id), bz)
+}
+
+func (k Keeper) GetGVGFamily(ctx sdk.Context, spID, familyID uint32) (*types.GlobalVirtualGroupFamily, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	var gvgFamily types.GlobalVirtualGroupFamily
+	bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+	if bz == nil {
+		return nil, false
+	}
+	k.cdc.MustUnmarshal(bz, &gvgFamily)
+	return &gvgFamily, true
+}
+
+func (k Keeper) GenerateOrSetLVGForBucket(ctx sdk.Context, bucketID math.Uint, gvgID uint32) {
+}
+
+func (k Keeper) GetOrCreateEmptyGVGFamily(ctx sdk.Context, familyID uint32, spID uint32) (*types.GlobalVirtualGroupFamily, error) {
+	store := ctx.KVStore(k.storeKey)
+	var gvgFamily types.GlobalVirtualGroupFamily
+	if familyID == types.NoSpecifiedFamilyId {
+		id := k.GenNextGVGFamilyID(ctx)
+		gvgFamily = types.GlobalVirtualGroupFamily{
+			Id:                    id,
+			VirtualPaymentAddress: k.DeriveVirtualPaymentAccount(types.GVGFamilyName, id).String(),
+		}
+		return &gvgFamily, nil
+	} else {
+		bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+		if bz == nil {
+			return nil, types.ErrGVGFamilyNotExist
+		}
+		k.cdc.MustUnmarshal(bz, &gvgFamily)
+
+		storeSize := k.GetStoreSizeOfFamily(ctx, gvgFamily)
+		if storeSize > k.MaxStoreSizePerFamily(ctx) {
+			return nil, types.ErrLimitationExceed.Wrapf("The storage size within the family exceeds the limit. Current: %d, now: %d", k.MaxStoreSizePerFamily(ctx), storeSize)
+		}
+		if k.MaxGlobalVirtualGroupNumPerFamily(ctx) < uint32(len(gvgFamily.GlobalVirtualGroupIds)) {
+			return nil, types.ErrLimitationExceed.Wrapf("The gvg number within the family exceeds the limit.")
+		}
+		return &gvgFamily, nil
+	}
+}
+
+func (k Keeper) DeriveVirtualPaymentAccount(groupType string, id uint32) sdk.AccAddress {
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, id)
+
+	return address.Module(types.ModuleName, append([]byte(groupType), b...))
+}
+
+func (k Keeper) GetAvailableStakingTokens(ctx sdk.Context, gvg *types.GlobalVirtualGroup) math.Int {
+	stakingPrice := k.GVGStakingPrice(ctx)
+
+	mustStakingTokens := stakingPrice.MulInt64(int64(gvg.StoredSize))
+
+	return gvg.TotalDeposit.Sub(mustStakingTokens.TruncateInt())
+}
+
+func (k Keeper) BindingObjectToGVG(ctx sdk.Context, bucketID math.Uint, primarySPID, familyID, gvgID uint32, payloadSize uint64) (*types.LocalVirtualGroup, error) {
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return nil, types.ErrGVGNotExist
+	}
+
+	gvgFamily, found := k.GetGVGFamily(ctx, primarySPID, familyID)
```

Comment:
Do we need to check th storage limit of the gvg family here?

Reply:
No need to check. When the store size of the family reaches the limitation, it cannot serve more buckets, and the original bucket can continue to create objects

Reply:
OK, I see.